packages/lit-dev-content/samples/tutorials/first-component/06/after/word-viewer.ts

```diff
@@ -0,0 +1,49 @@
+import { css, html, LitElement } from 'lit';
+import { customElement, property, state } from 'lit/decorators.js';
+
+@customElement('word-viewer')
+class WordViewer extends LitElement {
+  static styles = css`
+    :host {
+      background-color: white;
+      color: violet;
+      cursor: pointer;
+      display: block;
+    }
+    pre {
+      padding: 0.2em;
+    }
+  `;
+
+  @state() private playDirection = 1;
+  @state() private idx = 0;
+  @property() words: string = 'initial value';
+
+  private intervalTimer?: number;
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.intervalTimer = setInterval(this.tickToNextWord, 1000);
+  }
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    clearInterval(this.intervalTimer);
+    this.intervalTimer = undefined;
+  }
+
+  render() {
+    const splitWords = this.words.split('.');
+    const word = splitWords[((this.idx % splitWords.length) + splitWords.length) % splitWords.length];
```

Comment:
I wonder if it's better to keep this as `splitWords[this.idx]` and make the logic that updates `this.idx` handle the bounds. Making sure it's between 0 and length in `tickToNextWord`? As we build up to this point then, the reader would not have to touch this part, only the `tickToNextWord`. Oh, then we'd need to split it again to grab the length.. 🤔

Making the property be type Array makes the attribute look ugly. A custom converter could be used to keep the period-separated string to convert to an array of strings but that seems maybe extra for this tutorial.

Does Lit have something like React's `useEffect` where we can set a length state every time the `words` property changes? Even if there is, I don't know if that's the best way to go about it.

Also
`(this.idx + splitWords.length) % splitWords.length` works with a single modulo.

Reply:
Agree that there's a bit too much going on with the index math here. I wish I had put it in a separate function called `wrap` to remove the noise.

Main reason I'd want to keep it like this is to have it be extremely similar to the youtube video.

Regarding the modulo, a double modulo is required for handling large negative wrapping. I.e.:

If length is 5, then -1, -6, -11 should all equal `4`.

The first modulo converts all the negative numbers like -6 and -11 into -1.
Then -1 + 5 = 4.
The second modulo is for positive cases which would now be outside the range.

Reply:
Ah, right. With the wrapping happening here means `this.idx` can grow very large in the negative direction as well. Thanks for clarifying.

I don't necessarily agree that we should keep it like this just for the sake of congruency with the video. If there are improvements we can make, I think we should implement those in the tutorial. Having said that, none of the options I propose above I'd consider a strict improvement, so am okay with keeping it as is.

Reply:
That's a great point. Maybe it's worth splitting out the idx logic into a `wrap` function? It doesn't change much and the function name may also add some clarity to what is going on.

Then `const word = splitWords[((this.idx % splitWords.length) + splitWords.length) % splitWords.length];` could become:
`const word = splitWords[wrap(this.idx, splitWords.length)];

Reply:
In practice, I might do that, but for this tutorial I wonder if that's just extra hop for readers. How about just using a variable?
```suggestion
    const idx = ((this.idx % splitWords.length) + splitWords.length) % splitWords.length;
    const word = splitWords[idx];
```

Reply:
Agree. Adding this change.